design/kata-api-design.md

```diff
@@ -0,0 +1,112 @@
+# Kata API Design
+To fulfill the [kata design requirements](kata-design-requirements.md), and based on the disscusion on [Virtcontainers API extentions](https://docs.google.com/presentation/d/1dbGrD1h9cpuqAPooiEgtiwWDGCYhVPdatq7owsKHDEQ), the kata runtime library features following APIs:
+-  Sandbox based top API
+-  Storage and network hotplug API
+-  Plugin frameworks for external proprietary Kata runtime extensions
+-  Built-in shim and proxy types and capabilities
+
+## Sandbox Based API
+### Sandbox Management API
+|Name|Description|
+|---|---|
+|CreateSandbox(SandboxConfig)| Create, start a sandbox and return the sandbox structure.|
+|FetchSandbox(ID)| Connect to an existing sandbox and return the sandbox structure.|
+|ListSandbox()| List all existing standboxes with status. |
+
+### Sandbox Operation API
+|Name|Description|
+|---|---|
+|sandbox.Pause()| Pause the sandbox.|
+|sandbox.Resume()| Resume the paused sandbox.|
+|sandbox.Delete()| Destroy the sandbox and remove all persistent metadata.|
+|sandbox.Status()| Get the status of the sandbox and containers.|
+|sandbox.Monitor()| Return a context handler for caller to monitor sandbox callbacks such as error termination.|
+|sandbox.CreateContainer()| Create new container in the sandbox.|
+|sandbox.DeleteContainer()| Delete a container from the sandbox.|
+|sandbox.StartContainer()| Start a container in the sandbox.|
+|sandbox.StatusContainer()| Get the status of a container in the sandbox.|
+|sandbox.EnterContainer()| Run a new process in a container.|
+|sandbox.WaitProcess()| Wait on a process.|
+### Sandbox Hotplug API
+|Name|Description|
+|---|---|
+|sandbox.AddStorage()| Add new storage to the sandbox.|
+|sandbox.AddNetwork()| Add new nic to the sandbox.|
+|sandbox.ListNetwork()| List all nics and their configurations in the sandbox.|
+|sandbox.UpdateNetwork()| Update the configuration of an existing network interface.|
+|sandbox.UpdateRoute()| Update the sandbox route table, e.g. for portmapping support.|
+
+### Sandbox Relay API
+|Name|Description|
+|---|---|
+|sandbox.WinsizeProcess(containerID, processID, Height, Width)|Relay TTY resize request to a process.|
+|sandbox.SignalProcess(containerID, processID, signalID, signalALL)| Relay a signal to a process or all processes in a container.|
+|sandbox.IOStream(containerID, processID)| Relay a process stdio. Return stdin/stdout/stderr pipes to the process stdin/stdout/stderr streams.|
+
+## Plugin framework for external proprietary Kata runtime extensions
+### Hypervisor plugin
+
+TBD.
+### Metadata storage plugin
+The metadata storage plugin controls where sandbox metadata is saved.
+All metadata storage plugins must implement following API:
```

Comment:
Ok I need to understand why we need a plugin here. My understanding was that a plugin was useful in cases where we wanted proprietary implementations.
But if the point is to support a new type of storage, it should be directly implemented inside virtcontainers since the point of this library is to be reused by other projects, and we want to limit the potential duplication.

Last thing, you're defining some functions `Save/Load/Delete` but I'd like to understand why this needs to be defined through virtcontainers. Are you gonna make virtcontainers interacting with this ? If not, my understanding is that this could be directly implemented from the caller.

Reply:
Agree. Is this an API/Interface or a plugin? Same question for VM Factory plugin below.

Reply:
@sboeuf @amshinde please see the picture illustrating when these APIs are called **within** virtcontainers.